Impossible-Agent-Suite/LargeDepositWithdrawOnPairs/src/pairs.fetcher.spec.ts

```diff
@@ -4,86 +4,109 @@ import { createAddress } from "forta-agent-tools";
+      expect(await fetcher.isImpossiblePair(block, addr)).toStrictEqual(false);
     }
   });
 
-  it("should fetch the pairs", async () => {
-    initBlock(20);
+  it("should return false in calls to isImpossiblePair for address that doesn't have token1", async () => {
+    for(let [addr, block,,token0,] of CASES){
+      mockProvider
+        .addCallTo(addr, block, abi.PAIR, "token0", { inputs:[], outputs:[token0]});
+      // no mock initialization for token1 so calls to token1 should return undefined
+      expect(await fetcher.isImpossiblePair(block, addr)).toStrictEqual(false);
+    }
+  });
 
-    const pairs: string[] = await fetcher.getAllPairs(20);
-    expect(pairs).toStrictEqual(PAIRS);
+  it("should return false in calls to isImpossiblePair when the factory mismatch", async () => {
+    for(let [addr, block, factory, token0, token1] of CASES){
+      mockProvider
+        .addCallTo(addr, block, abi.PAIR, "token0", { inputs:[], outputs:[token0]})
+        .addCallTo(addr, block, abi.PAIR, "token1", { inputs:[], outputs:[token1]})
+        .addCallTo(
+          factory, block, abi.FACTORY, "getPair", 
+          { inputs:[token0, token1], outputs:[createAddress(`0xdead${block}`)] }
+        );
+      const fetcher: PairFetcher = new PairFetcher(factory, mockProvider as any);
+      expect(await fetcher.isImpossiblePair(block, addr)).toStrictEqual(false);
+    }
   });
 
-  it("should fetch the correct amount of pairs", async () => {
-    initBlock(101);
-    mockProvider.addCallTo(factory, 101, abi.FACTORY, "allPairsLength", {
-      inputs: [],
-      outputs: [2],
-    });
+  it("should return true in calls to isImpossiblePair in valid pairs", async () => {
+    for(let [pair, block, factory, token0, token1] of CASES){
+      mockProvider
+        .addCallTo(pair, block, abi.PAIR, "token0", { inputs:[], outputs:[token0]})
+        .addCallTo(pair, block, abi.PAIR, "token1", { inputs:[], outputs:[token1]})
+        .addCallTo(
+          factory, block, abi.FACTORY, "getPair", 
+          { inputs:[token0, token1], outputs:[pair] }
+        );
+      const fetcher: PairFetcher = new PairFetcher(factory, mockProvider as any);
+      expect(await fetcher.isImpossiblePair(block, pair)).toStrictEqual(true);
 
-    const pairs: string[] = await fetcher.getAllPairs(101);
-    expect(pairs).toStrictEqual(PAIRS.slice(0, 2));
+      // clear mock to use values in cache
+      mockProvider.clear();
+      expect(await fetcher.isImpossiblePair(block, pair)).toStrictEqual(true);
+    }
   });
 
   it("should fetch the reserves", async () => {
-    mockProvider.addCallTo(PAIRS[0], 30, abi.PAIR, "getReserves", {
+    mockProvider.addCallTo(CASES[0][0], 30, abi.PAIR, "getReserves", {
       inputs: [],
       outputs: [5, 10, 0],
     });
-    mockProvider.addCallTo(PAIRS[2], 53, abi.PAIR, "getReserves", {
+    mockProvider.addCallTo(CASES[2][0], 53, abi.PAIR, "getReserves", {
       inputs: [],
       outputs: [1, 3, 0],
     });
```

Comment:
Why there's a 3rd value in the output? `getReserves` function returns 2 values

Reply:
Yes `PairFetcher.getReserves` returns 2 values, but the contract signature returns a 3rd one that is the timestamp of the latest update. 
Check here the signature:
https://bscscan.com/address/0x438A242B9774c6D562d43fA2c1410dc9d9dE3281#code#L70

Reply:
I got it, thanks for the clarification